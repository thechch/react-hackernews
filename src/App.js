import React, {Component} from 'react';
import Search from './Search';
import Table from './Table';
import Button from './Button';
import './App.css';


const DEFAULT_QUERY = 'redux';
const DEFAULT_PAGE = 0;
const DEFAULT_HPP = '100';

const PATH_BASE = 'https://hn.algolia.com/api/v1';
const PATH_SEARCH = '/search';
const PARAM_SEARCH = 'query=';
const PARAM_PAGE = 'page=';
const PARAM_HPP = 'hitsPerPage=';

//var url = `${PATH_BASE}${PATH_SEARCH}?${PARAM_SEARCH}${DEFAULT_QUERY}&${PARAM_PAGE}${DEFAULT_PAGE}`;

class App extends Component {
    constructor(props) {
        super(props);

        this.state = {
            results: null,
            searchKey: '',
            searchTerm: DEFAULT_QUERY,
            totalPages: null,
        };

        this.setSearchTopStories = this.setSearchTopStories.bind(this);
        this.fetchSearchTopStories = this.fetchSearchTopStories.bind(this);
        this.onDismiss = this.onDismiss.bind(this);
        this.onSearchChange = this.onSearchChange.bind(this);
        this.onSearchSubmit = this.onSearchSubmit.bind(this);
    }

    onDismiss(id) {
        const isNotId = item => item.objectID !== id;
        const updatedList = this.state.result.hits.filter(isNotId);

        this.setState({
            result: {...this.state.result, hits: updatedList},
        });
    }

    onSearchChange(event) {
        this.setState({
            searchTerm: event.target.value
        });
    }

    onSearchSubmit(event) {
        const {searchTerm} = this.state;
        this.setState({ searchKey: searchTerm });
        this.fetchSearchTopStories(searchTerm, DEFAULT_PAGE);
        event.preventDefault();
    }

    setSearchTopStories(result) {
        const {hits, page, nbPages} = result;
        const {searchKey, results} = this.state;

        const oldHits = results && results[searchKey]
            ? results[searchKey].hits
            : [];

        const updatedHits = [
            ...oldHits,
            ...hits
        ];

        this.setState({
            ...results,
            [searchKey]: {hits: updatedHits, page},
            totalPages: nbPages,
        });
    }

    fetchSearchTopStories(searchTerm, page) {
        fetch(`${PATH_BASE}${PATH_SEARCH}?${PARAM_SEARCH}${searchTerm}&${PARAM_PAGE}${page}&${PARAM_HPP}${DEFAULT_HPP}`)
            .then(response => response.json())
            .then(result => this.setSearchTopStories(result));
    }

    componentDidMount() {
        const {searchTerm}  = this.state;
        this.setState( {searchKey: searchTerm});
        this.fetchSearchTopStories(searchTerm, DEFAULT_PAGE);
    }

    render() {
        const {result, searchTerm, totalPages} = this.state;
        const currentPage = (result && result.page) || 0;
        console.log(result);
        const modifier = totalPages ? currentPage + 1 === totalPages ? 0 : 1 : 0;

        return (
            <div className="page">
                <div className="interactions">
                    <Search
                        value={searchTerm}
                        onChange={this.onSearchChange}
                        onSubmit={this.onSearchSubmit}
                    >
                        Search
                    </Search>
                </div>
                {
                    result && <Table
                        list={result.hits}
                        onDismiss={this.onDismiss}
                    />
                }
                <div className="interactions">
                    <Button onClick={() => this.fetchSearchTopStories(searchTerm, currentPage + modifier)}>
                        More
                    </Button>
                </div>
            </div>
        );
    }
}

export default App;
