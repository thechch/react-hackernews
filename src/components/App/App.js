import React, { Component } from 'react';
import Search from '../Search/Search';
import Table from '../Table/Table';
import Button from '../Button/Button';
import './App.css';


const DEFAULT_QUERY = 'redux';
const DEFAULT_PAGE = 0;
const DEFAULT_HPP = '100';

const PATH_BASE = 'https://hn.algolia.com/api/v1';
const PATH_SEARCH = '/search';
const PARAM_SEARCH = 'query=';
const PARAM_PAGE = 'page=';
const PARAM_HPP = 'hitsPerPage=';

class App extends Component {
    constructor(props) {
        super(props);

        this.state = {
            results: null,
            searchKey: '',
            searchTerm: DEFAULT_QUERY,
            totalPages: null,
        };

        this.needsToSearchTopStories = this.needsToSearchTopStories.bind(this);
        this.setSearchTopStories = this.setSearchTopStories.bind(this);
        this.fetchSearchTopStories = this.fetchSearchTopStories.bind(this);
        this.onDismiss = this.onDismiss.bind(this);
        this.onSearchChange = this.onSearchChange.bind(this);
        this.onSearchSubmit = this.onSearchSubmit.bind(this);
    }
    componentDidMount() {
        const { searchTerm } = this.state;
        this.fetchSearchTopStories(searchTerm, DEFAULT_PAGE);
    }
    onSearchChange(event) {
        this.setState({
            searchTerm: event.target.value,
        });
    }
    onDismiss(id) {
        const { searchKey, results } = this.state;
        const { hits, page } = results[searchKey];

        const isNotId = item => item.objectID !== id;
        const updatedHits = hits.filter(isNotId);

        this.setState({
            results: {
                ...results,
                [searchKey]: { hits: updatedHits, page },
            },
        });
    }
    onSearchSubmit(event) {
        const { searchTerm } = this.state;
        this.setState({ searchKey: searchTerm });
        if (this.needsToSearchTopStories(searchTerm)) {
            this.fetchSearchTopStories(searchTerm, DEFAULT_PAGE);
        }
        event.preventDefault();
    }
    setSearchTopStories(result) {
        const { hits, page, nbPages } = result;
        const { searchKey, results } = this.state;

        const oldHits = results && results[searchKey]
            ? results[searchKey].hits
            : [];

        const updatedHits = [
            ...oldHits,
            ...hits,
        ];

        this.setState({
            results: {
                ...results,
                [searchKey]: { hits: updatedHits, page },
            },
            totalPages: nbPages,
        });
    }
    needsToSearchTopStories(searchTerm) {
        return !this.state.results[searchTerm];
    }
    fetchSearchTopStories(searchTerm, page) {
        this.setState({ searchKey: searchTerm });
        fetch(`${PATH_BASE}${PATH_SEARCH}?${PARAM_SEARCH}${searchTerm}&${PARAM_PAGE}${page}&${PARAM_HPP}${DEFAULT_HPP}`)
            .then(response => response.json())
            .then(result => this.setSearchTopStories(result));
    }
    render() {
        const {
            results,
            searchTerm,
            searchKey,
            totalPages,
        } = this.state;
        const page = (results && results[searchKey] && results[searchKey].page) || 0;
        const list = (results && results[searchKey] && results[searchKey].hits) || [];
        const modifier = totalPages && page + 1 !== totalPages
            ? 1
            : 0;
        return (
          <div className="page">
            <div className="interactions">
              <Search
                value={searchTerm}
                onChange={this.onSearchChange}
                onSubmit={this.onSearchSubmit}
              >
                        Search
                    </Search>
            </div>
            <Table
              list={list}
              onDismiss={this.onDismiss}
            />
            <div className="interactions">
              <Button onClick={() => this.fetchSearchTopStories(searchKey, page + modifier)}>
                        Give me more
              </Button>
            </div>
          </div>
        );
    }
}

export default App;

export {
    Button,
    Search,
    Table,
};
