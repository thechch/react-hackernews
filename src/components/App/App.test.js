import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import App from './App';

describe('App', () => {
    it('renders', () => {
        const div = document.createElement('div');
        ReactDOM.render(<App />, div);
    });

    test('snapshots', () => {
        const component = renderer.create(<App />);
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
});
