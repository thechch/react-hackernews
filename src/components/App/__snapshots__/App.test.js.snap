exports[`App snapshots 1`] = `
<div
  className="page">
  <div
    className="interactions">
    <form
      onSubmit={[Function]}>
      Search 
       
      <input
        onChange={[Function]}
        type="text"
        value="redux" />
      <button
        type="submit">
        Search
      </button>
    </form>
  </div>
  <div
    className="table" />
  <div
    className="interactions">
    <div
      className="spinner">
      <div
        className="bounce1" />
      <div
        className="bounce2" />
      <div
        className="bounce3" />
    </div>
  </div>
</div>
`;
